### PR TITLE
feat: support change loaders in extend

### PR DESCRIPTION
**Pull request by clarkdo**
TODO:
- [x] style loaders:
    css cssModules less sass scss stylus vueStyle
- [x] test cases
- [x] doc change https://github.com/nuxt/docs/pull/750

### CODE REVIEW

File: lib/builder/webpack/base.js [hunk 1]
```diff
@@ -276,7 +292,9 @@ export default class WebpackBaseConfig {
 
   customize(config) {
     if (typeof this.options.build.extend === 'function') {
-      const extendedConfig = this.options.build.extend.call(this.builder, config, this.nuxtEnv)
+      const extendedConfig = this.options.build.extend.call(
+        this.builder, config, { loaders: this.loaders, ...this.nuxtEnv }
```
Comment (pi0) [inline lib/builder/webpack/base.js]: Nice idea

**Review by pi0 — APPROVED** [review]

**Comment by ishitatsuyuki** [comment]
I recommend https://github.com/neutrinojs/webpack-chain instead of manually designing a configuration API.

**Comment by galvez** [comment]
@ishitatsuyuki I believe that's @clarkdo's ultimate goal. This is preliminary work in enabling the switch further down the road.

**Comment by ishitatsuyuki** [comment]
@galvez I don't see how this acts as preliminary work. `webpack-chain` uses a completely different (programmatic) API, but this PR adds a bunch of options described with an `Object`.

**Comment by clarkdo** [comment]
Thank you for the advice, yes, we have discussed that will support `webpack-chain`.

But this pull request is more involved loaders pre configuration, and `webpack-chain` is that how we config webpack and also include similarity as our current `extend` which is changing after configuration initialization.

With this pr, when starting to use webpack, we can config options easily:

```js
const name = fontUrlLoader
config.module
  .rule('font')
    .test(/\.(woff2?|eot|ttf|otf)(\?.*)?$/)
    .use(name)
      .loader('url-loader')
      .options(this.loaders[name]);
```

And about `webpakc-chain`, because it doesn't support parse from standard webpack config object into `chained` structure, so we may need to refactor whole webpack config classes, so we're refining the configuration classed to make it more compliant with the `webpack-chain` flow: `entry`, `plugin `, `devServer`, `optimizations`...

**Comment by ishitatsuyuki** [comment]
I see. Thanks for the explanation.

File: lib/builder/webpack/base.js [hunk 2]
```diff
@@ -313,6 +327,6 @@ export default class WebpackBaseConfig {
     }
 
     // Clone deep avoid leaking config between Client and Server
-    return _.cloneDeep(config)
+    return _.cloneDeep(this.customize(config))
```
Comment (pi0) [inline lib/builder/webpack/base.js]: Not related to this PR. but can we rename `customize` to `extendConfig`? @clarkdo

File: lib/common/options.js [hunk 3]
```diff
@@ -209,6 +209,22 @@ Options.from = function (_options) {
...
+  if (vueLoader.productionMode === undefined) {
+    vueLoader.productionMode = !options.dev
+  }
+  const styleLoaders = [
+    'css', 'cssModules', 'less',
+    'sass', 'scss', 'stylus', 'vueStyle'
+  ]
+  for (const name of styleLoaders) {
+    const loader = loaders[name]
+    if (loader) {
```
Comment (pi0) [inline lib/common/options.js]: Suggest changing to `if (loader && typeof loader.sourceMap === 'undefined')` so that user can override it

Comment (pi0) [reply lib/common/options.js]: I know this is nothing common but the way that we override it makes it impossible :)

**Review by pi0 — APPROVED** [review]

**Pull request merged**